sa/sa.go

```diff
@@ -713,39 +667,46 @@ func (ssa *SQLStorageAuthority) UpdatePendingAuthorization(ctx context.Context,
 		return
 	}
 
+	// If the provided authz isn't Status: pending that's a problem, return early.
 	if !statusIsPending(authz.Status) {
 		err = errors.New("Use FinalizeAuthorization() to update to a final status")
 		err = Rollback(tx, err)
 		return
 	}
 
-	if existingFinal(tx, authz.ID) {
-		err = errors.New("Cannot update a final authorization")
+	dbAuthz, table, err := getAuthz(tx, authz.ID)
+	if err != nil {
 		err = Rollback(tx, err)
 		return
 	}
 
-	if !existingPending(tx, authz.ID) {
-		err = errors.New("Requested authorization not found " + authz.ID)
+	// If the existing authz row isn't pending, we can't update it
+	if !statusIsPending(dbAuthz.Status) {
+		err = errors.New("Cannot update a non-pending authorization")
 		err = Rollback(tx, err)
 		return
 	}
 
-	var pa pendingauthzModel
-	err = tx.SelectOne(&pa, fmt.Sprintf("SELECT %s FROM pendingAuthorizations WHERE id = ?", pendingAuthzFields), authz.ID)
-	if err == sql.ErrNoRows {
-		return Rollback(tx, fmt.Errorf("No pending authorization with ID %s", authz.ID))
-	}
-	if err != nil {
-		return Rollback(tx, err)
+	var updateAuth interface{}
+	if table == "pendingAuthorizations" {
+		// If the authz came from the legacy pending table, use
+		// a `pendingAuthzModel` as the `updateAuth`.
+		updateAuth = &pendingauthzModel{Authorization: authz}
+	} else if table == "authz" {
+		// If the authz came from the authz table, use an authzModel
+		updateAuth = &authzModel{Authorization: authz}
+	} else {
+		// Should never happen - we only have two fixed authz tables!
+		err = errors.New("Internal error. Table mismatch updating authz")
```

Comment:
`Table mismatch` here seems slightly misleading, perhaps `Unknown table` would be more appropriate?